gems/quilt_rails/README.md

```diff
@@ -215,18 +217,205 @@ class ReactController < ApplicationController
+StatsD.distribution('navigation_usable', 23924, ['browser_connection_type:3g'])
+```
+
+##### Customizing `process_report` with a block
+
+The behaviour of `process_report` can be customized by manipulating the `Quilt::Performance::Client` instance yielded into it's implicit block parameter.
+
+```ruby
+process_report do |client|
+  # client.on_distribution do ....
+end
+```
+
+#### Client
+
+The `Quilt::Performance::Client` class is yielded into the block parameter for `process_report`, and is the primary API for customizing what metrics are sent for a given POST.
+
+##### Client#on_distribution
+
+The `on_distribution` method takes a block which is run for each distribution (including custom ones) sent during `process_report`. It can be used to easily add logging or other side-effects to your measurements.
+
+```ruby
+client.on_distribution do |metric_name, value, tags|
+  Rails.logger.debug "#{metric_name}: #{value}, tags: #{tags}"
+end
+```
+
+##### Client#on_distribution
+
+The `on_distribution` method takes a block which is run for each distribution (including custom ones) sent during `process_report`. It can be used to easily add logging or other side-effects to your measurements.
+
+```ruby
+client.on_distribution do |metric_name, value, tags|
+  Rails.logger.debug "#{metric_name}: #{value}, tags: #{tags}"
+end
+```
+
+##### Client#on_report
+
+The `on_report` method takes a block which is run once per report POSTed to the controller. It can be used to perform side-effects based on the entire `Report`, including sending custom `distribution`s.
+
+```ruby
+client.on_report do |report, tags|
+  client.distribution('some_metric', some_computation(report), tags)
+end
+```
+
+##### Client#on_navigation
+
+The `on_report` method takes a block which is run once per navigation in a given `Report`. It can be used to perform side-effects based on a given `Navigation`, including sending custom `distribution`s.
+
+```ruby
+client.on_navigation do |navigation, tags|
+  weight = navigation.events_with_size.reduce(0) do |total, event|
+    total + event.size
+  end
+
+  client.distribution('navigation_total_resource_weight', weight, tags)
+end
+```
+
+##### Client#customize_tags
+
+The `customize_tags` method allows custom tags to be computed to include in every distribution for a given POST to the controller.
+
+```ruby
+client.customize_tags do |metrics_body, tags|
+  # pathname will be inluded on *every* distribution for this report
+  tags[:pathname] = metrics_body.pathname
+end
+```
+
+##### Client#customize_navigation_tags
+
+The `customize_navigation_tags` method allows custom tags to be computed for each individual navigation described by the POST body, and included on the `distribution` calls for that navigation.
+
+```ruby
+client.customize_navigation_tags do |navigation, tags|
+  # target will be included only for the distributions about this navigation
+  tags[:navigation_target] = navigation.target
+end
+```
+
+#### Report
```

Comment:
It's kind of weird to me to have `Report` be a meaningful thing since it represents a collection of navigations/ events of unknown size/ distribution :/

Reply:
`Report` represents a specifically a report sent from the client. We even call the client-side component `<PerformanceReport />`. I'm not sure what else we'd name the class?

Reply:
Well `PerformanceReport` is a long-lived thing that sends all events to the server. I am just talking about how we interact with that on the server, and what I am saying in that instance is that anything that represents the "report" as being significant is weird (because it is literally just a throttled grouping of random events from the client).

Reply:
I'm not sure what you mean by significant. This literally says:

> describes timing, size information, and metadata about a series of events and navigations on the front-end of a web application

This thing, whatever it is called, is what the user of the library will need to understand the shape of to be able to write custom metrics code using `on_report`. If we called it `on_metrics_body` and named this `metrics_body` like we did in `koa-performance` I think that actually makes things less clear.

It will always be "significant" in the sense that it is part of the API consumers use to customize handling of the incoming requests. This is the same as it is in `koa-metrics` but with a name that feels more accurate / more obviously connected to the component that sends it on the react side.

Reply:
But why do they do anything on a report? Sure, they can add stuff on navigations, and they can add stuff on events, but a report is basically just a big array. Why does that need to be exposed to the consumer? Put another way, if the `PerformanceReport` were not batching (it was just sending events as it gets them), would this construct make any sense?

Reply:
Report also contains other top-level info though, such as extra metadata about the browser and it's internet connection. If you want to generate tags off of those we'd either need to thread that through to the navigation (which I guess is reasonable) or expose the report itself.

Reply:
🤔 I'm going to explore just passing the `connection` down to `navigation` and see how that feels

Reply:
tldr: it feels awful